File: android/app/src/main/java/ae/findmybay/data/repo/VendorRepository.kt
package ae.findmybay.data.repo

import ae.findmybay.data.api.VendorApi
import ae.findmybay.domain.model.AvailabilityResponse
import ae.findmybay.domain.model.Bay
import ae.findmybay.domain.model.Service
import ae.findmybay.domain.model.Slot
import ae.findmybay.domain.model.Vendor
import ae.findmybay.domain.model.VendorDetail
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class VendorRepository @Inject constructor(
    private val api: VendorApi,
) {
    suspend fun nearby(lat: Double, lng: Double, radiusMeters: Int = 15_000): List<Vendor> =
        api.nearby(lat = lat, lng = lng, radiusMeters = radiusMeters).items.map {
            Vendor(
                id = it.id,
                brandName = it.brandName,
                city = it.city,
                emirate = it.emirate,
                lat = it.lat,
                lng = it.lng,
                distanceMeters = it.distanceMeters,
                freeBays = it.freeBays,
                rating = it.rating,
                priceFromAed = it.priceFromAed,
                logoUrl = it.logoUrl,
            )
        }

    suspend fun detail(id: String): VendorDetail = api.detail(id).let {
        VendorDetail(
            id = it.id,
            brandName = it.brandName,
            city = it.city,
            emirate = it.emirate,
            addressLine = it.addressLine,
            lat = it.lat,
            lng = it.lng,
            rating = it.rating,
            priceFromAed = it.priceFromAed,
            logoUrl = it.logoUrl,
            services = it.services.map { s ->
                Service(s.id, s.name, s.durationMin, s.priceAed, s.vatInclusive)
            },
            bays = it.bays.map { b -> Bay(b.id, b.name, b.bayType, b.status) },
        )
    }

    suspend fun availability(vendorId: String, serviceId: String, date: String): AvailabilityResponse =
        api.availability(vendorId, serviceId, date).let {
            AvailabilityResponse(
                serviceId = it.serviceId,
                serviceName = it.serviceName,
                durationMin = it.durationMin,
                date = it.date,
                slots = it.slots.map { s -> Slot(s.startsAt, s.endsAt, s.available) },
            )
        }
}

File: ios/FindMyBay/Data/Repo/VendorRepository.swift
// VendorRepository.swift
//
// Maps `VendorAPI` DTOs into pure domain models. Mirrors `VendorRepository.kt`.

import Foundation

actor VendorRepository {
    private let api: VendorAPI

    init(api: VendorAPI) {
        self.api = api
    }

    func nearby(lat: Double, lng: Double, radiusMeters: Int = 15_000) async throws -> [Vendor] {
        let resp = try await api.nearby(lat: lat, lng: lng, radiusMeters: radiusMeters, serviceId: nil)
        return resp.items.map(Self.toDomain)
    }

    func detail(id: String) async throws -> VendorDetail {
        let dto = try await api.detail(id: id)
        return VendorDetail(
            id: dto.id, brandName: dto.brandName, city: dto.city, emirate: dto.emirate,
            addressLine: dto.addressLine, lat: dto.lat, lng: dto.lng,
            rating: dto.rating, priceFromAed: dto.priceFromAed, logoUrl: dto.logoUrl,
            services: dto.services.map {
                Service(id: $0.id, name: $0.name, durationMin: $0.durationMin,
                        priceAed: $0.priceAed, vatInclusive: $0.vatInclusive)
            },
            bays: dto.bays.map { Bay(id: $0.id, name: $0.name, bayType: $0.bayType, status: $0.status) }
        )
    }

    func availability(vendorId: String, serviceId: String, date: String) async throws -> AvailabilityResponse {
        let dto = try await api.availability(id: vendorId, serviceId: serviceId, date: date)
        return AvailabilityResponse(
            serviceId: dto.serviceId, serviceName: dto.serviceName,
            durationMin: dto.durationMin, date: dto.date,
            slots: dto.slots.map { Slot(startsAt: $0.startsAt, endsAt: $0.endsAt, available: $0.available) }
        )
    }

    // MARK: -

    private static func toDomain(_ d: VendorDTO) -> Vendor {
        Vendor(
            id: d.id, brandName: d.brandName, city: d.city, emirate: d.emirate,
            lat: d.lat, lng: d.lng, distanceMeters: d.distanceMeters,
            freeBays: d.freeBays, rating: d.rating,
            priceFromAed: d.priceFromAed, logoUrl: d.logoUrl
        )
    }
}
